presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizationManager.java

```diff
@@ -0,0 +1,314 @@
+
+    private final ScheduledExecutorService discoveryService = newScheduledThreadPool(1, daemonThreadsNamed("shard-organization-discovery"));
+    private final AtomicBoolean started = new AtomicBoolean();
+    private final AtomicBoolean shutdown = new AtomicBoolean();
+
+    private final IDBI dbi;
+    private final MetadataDao metadataDao;
+    private final ShardOrganizerDao shardOrganizerDao;
+    private final ShardManager shardManager;
+
+    private final boolean enabled;
+    private final long organizationIntervalSeconds;
+
+    private final String currentNodeIdentifier;
+    private final ShardOrganizer organizer;
+
+    private final Set<Long> tablesInProgress = newConcurrentHashSet();
+
+    @Inject
+    public ShardOrganizationManager(
+            @ForMetadata IDBI dbi,
+            NodeManager nodeManager,
+            ShardManager shardManager,
+            ShardOrganizer organizer,
+            StorageManagerConfig config)
+    {
+        this(dbi,
+                nodeManager.getCurrentNode().getNodeIdentifier(),
+                shardManager,
+                organizer,
+                config.isOrganizationEnabled(),
+                config.getOrganizationInterval());
+    }
+
+    public ShardOrganizationManager(
+            IDBI dbi,
+            String currentNodeIdentifier,
+            ShardManager shardManager,
+            ShardOrganizer organizer,
+            boolean enabled,
+            Duration organizationInterval)
+    {
+        this.dbi = requireNonNull(dbi, "dbi is null");
+        this.metadataDao = onDemandDao(dbi, MetadataDao.class);
+        this.shardOrganizerDao = onDemandDao(dbi, ShardOrganizerDao.class);
+
+        this.organizer = requireNonNull(organizer, "organizer is null");
+        this.shardManager = requireNonNull(shardManager, "shardManager is null");
+        this.currentNodeIdentifier = requireNonNull(currentNodeIdentifier, "currentNodeIdentifier is null");
+
+        this.enabled = enabled;
+
+        requireNonNull(organizationInterval, "organizationInterval is null");
+        this.organizationIntervalSeconds = max(1, organizationInterval.roundTo(SECONDS));
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (!enabled) {
+            return;
+        }
+
+        if (!started.getAndSet(true)) {
+            startDiscovery();
+        }
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        shutdown.set(true);
+        discoveryService.shutdownNow();
+    }
+
+    private void startDiscovery()
+    {
+        discoveryService.scheduleWithFixedDelay(() -> {
+            if (shutdown.get()) {
+                return;
+            }
+
+            try {
+                // jitter to avoid overloading database
+                SECONDS.sleep(ThreadLocalRandom.current().nextLong(1, 5 * 60));
+
+                log.info("Running shard organizer...");
+                Set<Long> tableIds = discoverAndInitializeTablesToOrganize();
```

Comment:
Inline this variable. If you need to set a breakpoint, it's easy to do that on the `return` line of the method